src/main/kotlin/io/github/nomisrev/validation.kt

```diff
@@ -51,88 +46,101 @@ data class InvalidEmailOrPassword(override val errors: NonEmptyList<String>) : I
   override val field: String = "email or password"
 }
 
-fun Login.validate(): Validated<IncorrectInput, Login> =
-  email.validEmail().zip(password.validPassword(), ::Login).mapLeft(::IncorrectInput)
+fun Login.validate(): Either<IncorrectInput, Login> =
+  zipOrAccumulate(email.validEmail(), password.validPassword(), ::Login).mapLeft(::IncorrectInput)
 
-fun RegisterUser.validate(): Validated<IncorrectInput, RegisterUser> =
-  username
-    .validUsername()
-    .zip(email.validEmail(), password.validPassword(), ::RegisterUser)
+fun RegisterUser.validate(): Either<IncorrectInput, RegisterUser> =
+  zipOrAccumulate(
+      username.validUsername(),
+      email.validEmail(),
+      password.validPassword(),
+      ::RegisterUser
+    )
     .mapLeft(::IncorrectInput)
 
-fun Update.validate(): Validated<IncorrectInput, Update> =
-  username
-    .traverse(String::validUsername)
-    .zip(email.traverse(String::validEmail), password.traverse(String::validPassword)) {
-      username,
-      email,
-      password ->
+fun Update.validate(): Either<IncorrectInput, Update> =
+  zipOrAccumulate(
+      username.mapOrAccumulate(String::validUsername),
+      email.mapOrAccumulate(String::validEmail),
+      password.mapOrAccumulate(String::validPassword)
+    ) { username, email, password ->
       Update(userId, username, email, password, bio, image)
     }
     .mapLeft(::IncorrectInput)
 
-private fun <E, A, B> A?.traverse(f: (A) -> ValidatedNel<E, B>): ValidatedNel<E, B?> =
-  this?.let(f) ?: null.validNel()
+private fun <E, A, B> A?.mapOrAccumulate(f: (A) -> EitherNel<E, B>): EitherNel<E, B?> =
+  this?.let(f) ?: null.right()
 
 private const val MIN_PASSWORD_LENGTH = 8
 private const val MAX_PASSWORD_LENGTH = 100
 private const val MAX_EMAIL_LENGTH = 350
 private const val MIN_USERNAME_LENGTH = 1
 private const val MAX_USERNAME_LENGTH = 25
 
-private fun String.validPassword(): ValidatedNel<InvalidPassword, String> =
-  notBlank()
-    .zip(minSize(MIN_PASSWORD_LENGTH), maxSize(MAX_PASSWORD_LENGTH)) { a, _, _ -> a }
+private fun String.validPassword(): EitherNel<InvalidPassword, String> =
+  zipOrAccumulate(notBlank(), minSize(MIN_PASSWORD_LENGTH), maxSize(MAX_PASSWORD_LENGTH)) { a, _, _
+      ->
+      a
+    }
```

Comment:
Yikes, is this the formatting coming from spotless 😱 If yes, can we change this to:

```suggestion
  zipOrAccumulate(
    notBlank(),
    minSize(MIN_PASSWORD_LENGTH),
    maxSize(MAX_PASSWORD_LENGTH)
  ) { a, _, _ ->  a }
```

Reply:
Yes, and I think it's absolutely terrible, too. :-)
I think we will need to ignore spotless somehow, as the code looked like you propose before and spotless didn't let that pass. Or maybe override some rules somewhere? I never used spotless before, what would be your preferred way of solving this?

Reply:
`Caused by: org.gradle.api.GradleException: The following files had format violations:
    src/main/kotlin/io/github/nomisrev/validation.kt
        @@ -56,21 +56,21 @@
         
         fun·RegisterUser.validate():·Either<IncorrectInput,·RegisterUser>·=
         ··zipOrAccumulate(
        -····username.validUsername(),
        -····email.validEmail(),
        -····password.validPassword(),
        -····::RegisterUser
        -··)
        +······username.validUsername(),
        +······email.validEmail(),
        +······password.validPassword(),
        +······::RegisterUser
        +····)
         ····.mapLeft(::IncorrectInput)
         
         fun·Update.validate():·Either<IncorrectInput,·Update>·=
         ··zipOrAccumulate(
        -····username.mapOrAccumulate(String::validUsername),
        -····email.mapOrAccumulate(String::validEmail),
        -····password.mapOrAccumulate(String::validPassword)
        -··)·{·username,·email,·password·->
        -····Update(userId,·username,·email,·password,·bio,·image)
        -··}
        +······username.mapOrAccumulate(String::validUsername),
        +······email.mapOrAccumulate(String::validEmail),
        +······password.mapOrAccumulate(String::validPassword)
        +····)·{·username,·email,·password·->
        +······Update(userId,·username,·email,·password,·bio,·image)
        +····}
         ····.mapLeft(::IncorrectInput)
         
         private·fun·<E,·A,·B>·A?.mapOrAccumulate(f:·(A)·->·EitherNel<E,·B>):·EitherNel<E,·B?>·=
        @@ -83,30 +83,33 @@
         private·const·val·MAX_USERNAME_LENGTH·=·25
         
         private·fun·String.validPassword():·EitherNel<InvalidPassword,·String>·=
        -··zipOrAccumulate(
        -····notBlank(),
        -····minSize(MIN_PASSWORD_LENGTH),
        -····maxSize(MAX_PASSWORD_LENGTH)
        -··)·{·a,·_,·_·->·a·}
        +··zipOrAccumulate(notBlank(),·minSize(MIN_PASSWORD_LENGTH),·maxSize(MAX_PASSWORD_LENGTH))·{·a,·_,·_
        +······->
        +······a
        +····}
         ····.mapLeft(toInvalidField(::InvalidPassword))
         
         private·fun·String.validEmail():·EitherNel<InvalidEmail,·String>·{
    ... (31 more lines that didn't fit)`

Reply:
I'll give it a try

Reply:
It's actually KtFmt, Spotless is just a Gradle wrapper, https://github.com/facebook/ktfmt
So not sure an issue at Spotless will help us, I checked and it's indeed forcing us that weird style.

Let's leave it for now, it seems we cannot opt a region out.